com.unity.ml-agents.extensions/Runtime/Sensors/BoxOverlapChecker.cs

```diff
@@ -0,0 +1,254 @@
+
+                ParseCollidersClosest(m_ColliderBuffer, numFound, cellIndex, cellCenter, GridOverlapDetectedDebug);
+            }
+        }
+
+        /// <summary>
+        /// Parses the array of colliders found within a cell. Finds the closest gameobject to the agent root reference within the cell
+        /// </summary>
+        void ParseCollidersClosest(Collider[] foundColliders, int numFound, int cellIndex, Vector3 cellCenter, Action<GameObject, int> detectedAction)
+        {
+            GameObject closestColliderGo = null;
+            var minDistanceSquared = float.MaxValue;
+
+            for (var i = 0; i < numFound; i++)
+            {
+                var currentColliderGo = foundColliders[i].gameObject;
+
+                // Continue if the current collider go is the root reference
+                if (ReferenceEquals(currentColliderGo, m_RootReference))
+                {
+                    continue;
+                }
+
+                var closestColliderPoint = foundColliders[i].ClosestPointOnBounds(cellCenter);
+                var currentDistanceSquared = (closestColliderPoint - m_RootReference.transform.position).sqrMagnitude;
+
+                if (currentDistanceSquared >= minDistanceSquared)
+                {
+                    continue;
+                }
+
+                // Checks if our colliders contain a detectable object
+                var index = -1;
+                for (var ii = 0; ii < m_DetectableTags.Length; ii++)
+                {
+                    if (currentColliderGo.CompareTag(m_DetectableTags[ii]))
+                    {
+                        index = ii;
+                        break;
+                    }
+                }
+                if (index > -1 && currentDistanceSquared < minDistanceSquared)
+                {
+                    minDistanceSquared = currentDistanceSquared;
+                    closestColliderGo = currentColliderGo;
+                }
+            }
+
+            if (!ReferenceEquals(closestColliderGo, null))
+            {
+                detectedAction.Invoke(closestColliderGo, cellIndex);
+            }
+        }
+
+        /// <summary>
+        /// Parses all colliders in the array of colliders found within a cell.
+        /// </summary>
+        void ParseCollidersAll(Collider[] foundColliders, int numFound, int cellIndex, Vector3 cellCenter, Action<GameObject, int> detectedAction)
+        {
+            for (int i = 0; i < numFound; i++)
+            {
+                var currentColliderGo = foundColliders[i].gameObject;
+                if (!ReferenceEquals(currentColliderGo, m_RootReference))
+                {
+                    detectedAction.Invoke(currentColliderGo, cellIndex);
+                }
+            }
+        }
+
+        internal void RegisterSensor(GridSensorBase sensor)
+        {
+            if (sensor.ProcessAllCollidersInCell())
```

Comment:
nit: instead of `bool ProcessAllCollidersInCell()` and two separate registration methods, could you make the GridSensorBase return an enum (ProcessAllColliders, ProcessClosestCollider, DebugColliders) ? I think that would simplify this a bit.

Reply:
I made two separate registration methods because I don't want to expose `DebugColliders` since user should never use that.
I can make enum for `ProcessAllColliders` and `ProcessClosestCollider`.